src/cc/frontends/clang/b_frontend_action.cc

```diff
@@ -269,7 +269,37 @@ bool ProbeVisitor::VisitVarDecl(VarDecl *D) {
   }
   return true;
 }
+
+bool ProbeVisitor::IsAddrOfExpr(Expr *E) {
+  if (UnaryOperator *UnaryExpr = dyn_cast<UnaryOperator>(E)) {
+    if (UnaryExpr->getOpcode() == UO_AddrOf)
+      return true;
+  } else if (ExplicitCastExpr *ExplicitExpr = dyn_cast<ExplicitCastExpr>(E)) {
+    return IsAddrOfExpr(ExplicitExpr->getSubExpr());
+  } else if (ImplicitCastExpr *ImplicitExpr = dyn_cast<ImplicitCastExpr>(E)) {
+    return IsAddrOfExpr(ImplicitExpr->getSubExpr());
+  } else if (CStyleCastExpr *CStyleExpr = dyn_cast<CStyleCastExpr>(E)) {
+    return IsAddrOfExpr(CStyleExpr->getSubExpr());
+  }
+  return false;
+}
+
 bool ProbeVisitor::VisitCallExpr(CallExpr *Call) {
+  // Skip bpf_probe_read for the third argument if it is an AddrOf.
+  if (VarDecl *V = dyn_cast<VarDecl>(Call->getCalleeDecl())) {
+    if (V->getName() == "bpf_probe_read") {
+      unsigned i = 0;
+      // check the third argument which is the src pointer.
+      for (auto arg : Call->arguments()) {
+        i++;
+        if (i != 3)
+          continue;
+        return !IsAddrOfExpr(arg);
```

Comment:
Wouldn't `return !isAddrOfExpr(Call->getArg(2))` be the same (after a check on `Call->getNumArgs()` maybe)?

Reply:
This works too. Thanks for all suggestions. This simplified the code a lot!